aptos/api/aux-ts/src/indexer/index.ts

```diff
@@ -86,188 +87,161 @@ async function publishAmmEvents() {
+  const key = `${baseCoinType}-${quoteCoinType}`;
+  const fills = _(await market.fills())
+    .filter((fill) => fill.sequenceNumber.toNumber() >= keyToCursor[key]!)
+    .value();
+  const cursor = fills[fills.length - 1]?.sequenceNumber.toNumber();
 
-  const now = Date.now();
-  fills = _.dropWhile(
-    fills,
-    (fill) => fill.timestamp < now - resolutionSeconds(resolution)
+  await publishFills(
+    fills.map((fill) =>
+      orderEventToOrder(fill, market.baseCoinInfo, market.quoteCoinInfo)
+    ),
+    key
   );
 
-  while (true) {
-    // pagination logic
-    // start: 0, limit: 100
-    // server will return starting from lowest seq num
-    // if there are multiple pages, keep paging until the cursor overlaps
-    // only process the records within a page that are > cursor
-    // move the cursor forward after processing
-    const moreFills = await market.fills({ start: cursor, limit: 100 });
-    const fillsSince =
-      cursor === 0
-        ? moreFills
-        : _.dropWhile(
-            moreFills,
-            (fill) => fill.sequenceNumber.toNumber() <= cursor
-          );
-    fills = fills.concat(fillsSince.map(convert));
-    cursor =
-      fillsSince[fillsSince.length - 1]?.sequenceNumber.toNumber() ?? cursor;
-    if (fillsSince.length === 0 || fillsSince.length < moreFills.length) {
-      break;
-    }
+  if (!_.isUndefined(cursor)) {
+    keyToCursor[key] = cursor;
+    // publish bars
+    await Promise.all(
+      RESOLUTIONS.map((resolution) => publishBar(resolution, key))
+    );
   }
+}
+
+function publishFills(fills: Order[], key: string) {
+  return Promise.all([
+    fills.forEach(async (fill) => await redisPubSub.publish("TRADE", fill)),
+    fills.forEach(
+      async (fill) => await redisPubSub.publish("LAST_TRADE_PRICE", fill.price)
+    ),
+    redisClient.rPush(
+      `${key}-fills-24h`,
+      fills.map((fill) =>
+        JSON.stringify({
+          time: fill.time,
+          price: fill.price,
+          quantity: fill.quantity,
+        })
+      )
+    ),
+    Promise.all(
+      RESOLUTIONS.map((resolution) => {
+        redisClient.rPush(
+          `${key}-fills-${resolution}`,
+          fills.map((fill) =>
+            JSON.stringify({
+              time: fill.time,
+              price: fill.price,
+              quantity: fill.quantity,
+            })
+          )
+        );
+      })
+    ),
+  ]);
+}
+
+async function publishBar(resolution: Resolution, key: string): Promise<void> {
+  key = `${key}-fills-${resolution}`;
+  const now = Date.now();
+  const delta = now % (resolutionToSeconds(resolution) * 1000);
+  const time = now - delta;
+
+  const len = await redisClient.lLen(key);
+  const rawFills = await redisClient.lPopCount(key, len) ?? [];
+  const fills: Order[] = rawFills
+    .map((s) => JSON.parse(s))
+    .filter((fill: Order) => Number(fill.time) >= time);
 
+  if (_.isEmpty(fills)) {
+    await redisPubSub.publish("BAR", { resolution, time, ohlcv: null });
```

Comment:
Why do we need to publish here (as opposed to no op)?

Reply:
good q, i was considering the same and it's just an explicit null for that time vs. an implicit time gap